runtime/src/main/java/org/corfudb/runtime/clients/ClientResponseHandler.java

```diff
@@ -0,0 +1,235 @@
+    public interface Handler {
+        Object handle(ResponseMsg msg, ChannelHandlerContext ctx, IClientRouter r) throws Exception;
+    }
+
+    /**
+     * The handler map for normal ResponseMsg.
+     */
+    private final Map<PayloadCase, Handler> handlerMap;
+
+    /**
+     * The handler map for ServerErrorMsg.
+     */
+    private final Map<ErrorCase, Handler> errorHandlerMap;
+
+    /**
+     * The client.
+     */
+    private final IClient client;
+
+    /**
+     * Construct a new instance of ClientResponseHandler.
+     *
+     * @param client The client that this ClientResponseHandler will register.
+     */
+    public ClientResponseHandler(IClient client) {
+        this.client = client;
+        this.handlerMap = new ConcurrentHashMap<>();
+        this.errorHandlerMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Add a handler to this ClientResponseHandler.
+     *
+     * @param payloadCase Response payload type.
+     * @param handler The handler itself.
+     * @return This handler, to support chaining.
+     */
+    @Deprecated
+    public ClientResponseHandler addHandler(PayloadCase payloadCase,
+                                            ClientResponseHandler.Handler handler) {
+        if (!payloadCase.equals(PayloadCase.SERVER_ERROR)) {
+            handlerMap.put(payloadCase, handler);
+        }
+        return this;
+    }
+
+    /**
+     * Handle an incoming Response from server.
+     *
+     * @param response The Response to handle.
+     * @param ctx The channel handler context.
+     * @return True if the message was handled successfully.
+     */
+    public boolean handle(ResponseMsg response, ChannelHandlerContext ctx) {
+        IClientRouter router = client.getRouter();
+        long requestId = response.getHeader().getRequestId();
+        PayloadCase payloadCase = response.getPayload().getPayloadCase();
+
+        if (payloadCase.equals(PayloadCase.SERVER_ERROR)) {
+            ErrorCase errorCase = response.getPayload().getServerError().getErrorCase();
+            if (errorHandlerMap.containsKey(errorCase)) {
+                try{
+                    // TODO: should we remove this clause?
+                    Object ret = errorHandlerMap.get(errorCase).handle(response, ctx, router);
+                    if (ret != null) {
+                        router.completeRequest(requestId, ret);
+                    }
+                } catch (Throwable e) {
+                    router.completeExceptionally(requestId, e);
+                }
+            }
+        } else if (handlerMap.containsKey(payloadCase)) {
+            try {
+                Object ret = handlerMap.get(payloadCase).handle(response, ctx, router);
+                if (ret != null) {
+                    router.completeRequest(requestId, ret);
+                }
+            } catch (Throwable e) {
+                // TODO: same reason, should we remove this clause?
+                router.completeExceptionally(requestId, e);
```

Comment:
I am not sure if this can be removed, since we have to complete the future exceptionally. Does this occur elsewhere if this is removed?